### PR TITLE
Added packaging, main classes and main methods.

### PR DESCRIPTION
**Pull request by zhenyaAreYouHere63**
Wrote the basis for further writing the code for the program.

### CODE REVIEW

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    public StudentServiceImpl(StudentList students) {
+        this.students = students;
+    }
+
+    @Override
+    public void createNewStudent(String firstName, String lastName, Faculty faculty, Group group,
+                                 Specialty specialty) {
+
+        Student newStudent = new Student(firstName, lastName, faculty, specialty, group);
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:29]: if u are interested: u can implement validation in a bit more complex way - to save all fields with errors to show them to user + error message for each field

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        students.students.add(newStudent);
+        System.out.println(newStudent + " successfully added");
+    }
+
+    @Override
+    public void addStudentToCourse(int studentId, SubjectName subjectName) {
+
+        if (idValidation(studentId)) {
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:37]: it makes sense to check negative case and throw an exception or write return to simplify code

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            Optional<Student> optionalStudent = students.students.stream().filter(student -> student.getId() == studentId)
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:38]: stidents.students looks not very good, pls rename something there

Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:38]: move this logic to StudentList class

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    .findFirst();
+
+            if (optionalStudent.isPresent()) {
+                studentSpecialty = optionalStudent.get().getSpecialty();
+                SubjectName compulsorySubjectOfStudent = studentSpecialty.getCompulsorySubject();
+                if (optionalStudent.get().getAttendedSubject().size() < 3 && !compulsorySubjectOfStudent.equals(subjectName)) {
+                    optionalStudent.get().getAttendedSubject().add(subjectName);
+                } else {
+                    if (optionalStudent.get().getAttendedSubject().size() >= 3) {
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:47]: optionalStudent.get() might be moved to a separate var

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    public void addStudentToCourse(int studentId, SubjectName subjectName) {
+
+        if (idValidation(studentId)) {
+            Optional<Student> optionalStudent = students.students.stream().filter(student -> student.getId() == studentId)
+                    .findFirst();
+
+            if (optionalStudent.isPresent()) {
+                studentSpecialty = optionalStudent.get().getSpecialty();
+                SubjectName compulsorySubjectOfStudent = studentSpecialty.getCompulsorySubject();
+                if (optionalStudent.get().getAttendedSubject().size() < 3 && !compulsorySubjectOfStudent.equals(subjectName)) {
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:44]: add a check if subject is learned at the speciality of the student

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 4 above)
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:47]: pls don't duplicate conditions
check firstly negative cases and throw exception and then there would be no need to check success case additioally

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    public void viewAllSubjects(int studentId) {
+        if (idValidation(studentId)) {
+            students.students.stream()
+                    .filter(student -> student.getId() == studentId)
+                    .findFirst()
+                    .ifPresentOrElse(student -> {
+                                System.out.println("Subjects of the student's choice");
+                                student.getAttendedSubject()
+                                        .stream().map(SubjectName::name)
+                                        .forEach(subject -> System.out.println(" -" + subject));
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:69]: get rid of 'system.out' in  this class

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    @Override
+    public void viewAllGrades(int studentId) {
+        if (idValidation(studentId)) {
+            Optional<Student> student = students.students.stream().filter(s -> s.getId() == studentId).findFirst();
+
+            if (student.isPresent()) {
+                student.get().getGrades().forEach((key, value) -> System.out.println("subject: " + key + ", grade: " + value));
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:85]: marks should be returned

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                List<SubjectName> attendedSubjects = foundStudent.getAttendedSubject();
+                Map<SubjectName, List<Integer>> studentGrades = foundStudent.getGrades();
+
+                long numberOfGrades = studentGrades.entrySet().stream()
+                        .filter(entry -> entry.getKey().equals(subjectName))
+                        .flatMapToInt(entry -> entry.getValue().stream().mapToInt(Integer::intValue))
+                        .count();
+
+                if (numberOfGrades > 0) {
+                    int totalGrades = studentGrades.entrySet().stream()
```
Comment (polinaucc) [inline src/main/java/com/study/service/impl/StudentServiceImpl.java:109]: in such an imol you are ging through grades list twice and it is not very optimal

**Review by polinaucc — COMMENTED** [review]
Part 1 'Students' functionality was disscused

File: src/main/java/com/study/dao/program/Specialty.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    RADIO_ELECTRONICS(Faculty.RADIOENGINEERING, SubjectName.PHYSICS),
+    RADIO_ENGINEERING(Faculty.RADIOENGINEERING, SubjectName.MATH),
+    PHILOSOPHY(Faculty.SOCIOLOGY, SubjectName.PHYSICAL_CULTURE),
+    SOCIOLOGY(Faculty.SOCIOLOGY, SubjectName.SCIENCE_OF_LAW),
+    INTERNATIONAL_ECONOMICS(Faculty.MANAGEMENT, SubjectName.INTERNATIONAL_COMMUNICATIONS),
+    MANAGEMENT_OF_ENTERPRISES(Faculty.MANAGEMENT, SubjectName.PRODUCTION_EQUIPMENT),
+    ECONOMIC_CYBERNETICS (Faculty.MANAGEMENT, SubjectName.CHEMISTRY);
+
+    private final Faculty faculty;
+    private final SubjectName compulsorySubject;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/program/Specialty.java:13]: yes, didn't understand the task, redid the code.

File: src/main/java/com/study/dao/program/Group.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    REE_11(Specialty.RADIO_ENGINEERING),
+    REL_11(Specialty.RADIO_ELECTRONICS),
+    SP_11(Specialty.PHILOSOPHY),
+    SS_11(Specialty.SOCIOLOGY),
+    MIE_11(Specialty.INTERNATIONAL_ECONOMICS),
+    MME_11(Specialty.MANAGEMENT_OF_ENTERPRISES),
+    MEC_11(Specialty.ECONOMIC_CYBERNETICS);
+    private final Specialty specialty;
+
+    Group(Specialty specialty) {
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/program/Group.java:14]: Deleted this class

File: src/main/java/com/study/dao/program/Specialty.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package com.study.dao.program;
+
+public enum Specialty {
+    RADIO_ELECTRONICS(Faculty.RADIOENGINEERING, SubjectName.PHYSICS),
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/program/Specialty.java:4]: Yes, this is possible, I did not consider the training accounting system in such depth.

File: src/main/java/com/study/dao/pojo/Student.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
+package com.study.dao.pojo;
+
+import com.study.dao.program.Faculty;
+import com.study.dao.program.Group;
+import com.study.dao.program.Specialty;
+import com.study.dao.program.SubjectName;
+import java.util.*;
+
+public class Student {
+    private static int lastId = 0;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/pojo/Student.java:10]: Used class Counter for this function in the rewritten code.

File: src/main/java/com/study/dao/pojo/Student.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    private int id;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/pojo/Student.java:11]: Used UUID for this situation.

File: src/main/java/com/study/dao/pojo/Student.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    private String firstName;
+    private String lastName;
+    private Faculty faculty;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/pojo/Student.java:14]: Changed the code, now these fields are needed there.

File: src/main/java/com/study/dao/pojo/Teacher.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
+package com.study.dao.pojo;
+
+import com.study.dao.program.SubjectName;
+import java.util.Objects;
+
+public class Teacher {
+
+    private static int lastId = 0;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/pojo/Teacher.java:8]: Added Counter class.

File: src/main/java/com/study/dao/program/SubjectName.java [hunk 16]
```diff
@@ -0,0 +1,6 @@
+package com.study.dao.program;
+
+public enum SubjectName {
+    PHYSICS, MATH, ENGLISH, CHEMISTRY, SCIENCE_OF_LAW, INTERNATIONAL_COMMUNICATIONS,
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/program/SubjectName.java:4]: Reworked the code to display items in the console.

File: src/main/java/com/study/dao/pojo/Student.java [hunk 17]
```diff
@@ -0,0 +1,119 @@
+package com.study.dao.pojo;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/pojo/Student.java:1]: Renamed this package to 'core'

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 3 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:38]: renamed to students.studentList

File: src/main/java/com/study/input/ClientStudentInput.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+    public void createNewStudentWithInput() {
+        System.out.println("Enter username");
+        String firstName = scanner.nextLine();
+
+        System.out.println("Enter lastname");
+        String lastName = scanner.nextLine();
+
+        System.out.println("Enter faculty");
+        Faculty faculty = Faculty.valueOf(scanner.nextLine());
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/input/ClientStudentInput.java:31]: Removed all enum classes.

File: src/main/java/com/study/service/impl/StudentServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        this.students = students;
+    }
+
+    @Override
+    public void createNewStudent(String firstName, String lastName, Faculty faculty, Group group,
+                                 Specialty specialty) {
+
+        Student newStudent = new Student(firstName, lastName, faculty, specialty, group);
+        students.students.add(newStudent);
+        System.out.println(newStudent + " successfully added");
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:31]: ok, moved to 'controller'

File: src/main/java/com/study/service/validation/IdValidation.java [hunk 20]
```diff
@@ -0,0 +1,23 @@
+package com.study.service.validation;
+
+import com.study.service.exception.IncorrectIdException;
+
+public interface IdValidation {
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/validation/IdValidation.java:5]: ok

File: src/main/java/com/study/service/validation/IdValidation.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+
+    default boolean idValidation(int studentId) {
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/validation/IdValidation.java:7]: ok

File: src/main/java/com/study/display/StudentMenu.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
+package com.study.display;
+
+import com.study.dao.collections.StudentList;
+import com.study.input.ClientStudentInput;
+import java.util.Scanner;
+
+public class StudentMenu {
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/display/StudentMenu.java:7]: Renamed these classes

File: src/main/java/com/study/dao/collections/StudentList.java [hunk 23]
```diff
@@ -0,0 +1,19 @@
+package com.study.dao.collections;
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/dao/collections/StudentList.java:1]: Renamed this package

File: src/main/java/com/study/input/ClientStudentInput.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+
+        System.out.println("Enter specialty");
+        Specialty specialty = Specialty.valueOf(scanner.nextLine());
+
+        if (!specialty.isMatchingFaculty(faculty)) {
```
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/input/ClientStudentInput.java:36]: Removed this checks

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 8 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:109]: Reworked this method

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 7 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:85]: Returned marks

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 4 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:47]: Fixed it

Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:47]: ok, moved to a separate variable

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 3 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java:38]: moved this logic to StudentList class

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 2 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java]: fixed processing exceptions, negative cases checked firstly

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 6 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java]: ridded 'system.out' in this class

File: src/main/java/com/study/service/validation/IdValidation.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/validation/IdValidation.java]: Applied the 'Chain of responsibility' pattern.

File: src/main/java/com/study/service/impl/StudentServiceImpl.java (hunk 5 above)
Comment (zhenyaAreYouHere63) [reply src/main/java/com/study/service/impl/StudentServiceImpl.java]: Added this check

**Pull request merged**